p2p/src/network/scheduler/p2p_network_scheduler_reducer.rs

```diff
@@ -1,54 +1,122 @@
-use std::collections::BTreeMap;
+use std::{collections::BTreeMap, sync::OnceLock};
 
-use openmina_core::error;
+use identify::P2pNetworkIdentifyStreamAction;
+use openmina_core::{bug_condition, error, warn, Substate};
+use request::{P2pNetworkKadRequestState, P2pNetworkKadRequestStatus};
+use token::{
+    AuthKind, DiscoveryAlgorithm, IdentifyAlgorithm, MuxKind, PingAlgorithm, Protocol,
+    RpcAlgorithm, StreamKind,
+};
+
+use crate::{
+    connection::{
+        incoming::P2pConnectionIncomingAction, outgoing::P2pConnectionOutgoingAction,
+        P2pConnectionState,
+    },
+    disconnection::P2pDisconnectionAction,
+    identify::P2pIdentifyAction,
+    P2pConfig, P2pPeerStatus, P2pState,
+};
 
 use super::{super::*, p2p_network_scheduler_state::P2pNetworkConnectionState, *};
 
 impl P2pNetworkSchedulerState {
-    pub fn reducer(&mut self, action: redux::ActionWithMeta<&P2pNetworkSchedulerAction>) {
+    pub fn reducer<Action, State>(
+        mut state_context: Substate<Action, State, Self>,
+        action: redux::ActionWithMeta<&P2pNetworkSchedulerAction>,
+    ) -> Result<(), String>
+    where
+        State: crate::P2pStateTrait,
+        Action: crate::P2pActionTrait<State>,
+    {
         let (action, meta) = action.split();
+        let scheduler_state = state_context.get_substate_mut()?;
+
         match action {
             P2pNetworkSchedulerAction::InterfaceDetected { ip, .. } => {
-                self.interfaces.insert(*ip);
+                scheduler_state.interfaces.insert(*ip);
+
+                let (dispatcher, state) = state_context.into_dispatcher_and_state();
+                let p2p_config: &P2pConfig = state.substate()?;
+
+                if let Some(port) = p2p_config.libp2p_port {
+                    dispatcher.push(P2pNetworkSchedulerEffectfulAction::InterfaceDetected {
+                        ip: *ip,
+                        port,
+                    });
+                }
+
+                Ok(())
             }
             P2pNetworkSchedulerAction::InterfaceExpired { ip, .. } => {
-                self.interfaces.remove(ip);
+                scheduler_state.interfaces.remove(ip);
+                Ok(())
             }
             P2pNetworkSchedulerAction::ListenerReady { listener } => {
-                self.listeners.insert(*listener);
+                scheduler_state.listeners.insert(*listener);
+                Ok(())
```

Comment:
do you need `Ok(())` everywhere? instead of for example a single `Ok(())` at the end, and any case that produces an error would short-circuit everything with a `return`.

Reply:
I don't need `Ok(())` for every action but I think it is better to know that nothing runs after that.